### PR TITLE
feat: benchmark setup

### PR DESCRIPTION
**Pull request by alexkeizer**
### Description:

Add a basic benchmark setup:
- `benchmark` command, which has a theorem-like syntax, but doesn't get added to the environment, and runs the proof 5 times, reporting both the individual and average runtime,
- similarly, a `logHeartbeats` tactic combinator, which measures the heartbeats used by an arbitrary tactic.
- Add a benchmark folder, with files for SHA512 runs with: 150, 225, and 400 steps (which are the numbers I've been reporting so far)
- Adds a `make benchmark` Make-target, to build the aforementioned files. Crucially, this target is *not* part of `make all`, for all our sanity's sake (running 400 steps of SHA512 for 5 times takes a while)

### Testing:

What tests have been run? Did `make all` succeed for your changes? Was
conformance testing successful on an Aarch64 machine? yes

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by alexkeizer** [comment]
@shigoel, @bollu: do we want to run benchmarks in CI? It could be nice to gather data automatically, but I worry a bit that the data might be unreliable (since we have little to no control over what else is running on the CI worker), and that it might slow down CI too much (just doing the 400 step benchmark alone already takes 10 minutes)

File: Benchmarks/Command.lean
Comment (bollu) [inline Benchmarks/Command.lean:16]:
```suggestion
  for _ in [0:n] do
```

why not use range syntax? :)

File: Benchmarks/Command.lean
Comment (bollu) [inline Benchmarks/Command.lean:32]:
```suggestion
  let avg := totalRunTime.toFloat / n.toFloat / 1000
  let geomean := (totalRunTime.toFloat.pow (1.0 / n.toFloat)) / 1000.0
  logInfo m!"\
{id}:
  average runtime over {n} runs:
    {avg}s
  geomean over {n} runs:
    {geomean}s

  indidividual runtimes:
    {runTimes}
"
```

File: Benchmarks/Command.lean
Comment (alexkeizer) [reply Benchmarks/Command.lean]: At the risk of losing my street cred: because I wasn't aware of range syntax 🫢

File: Benchmarks/Command.lean
Comment (bollu) [reply Benchmarks/Command.lean:32]: Ideally, we would also track variance, but I'm happy to do that in a follow up PR.

**Comment by shigoel** [comment]
@alexkeizer I like the idea of benchmarking automatically in CI (even with all the caveats that we can't control what else is running on the worker). Maybe we can separate the benchmarking job from others so that folks don't have to wait for the benchmarking job to finish before hitting merge, etc. if they don't want to.

**Comment by shigoel** [comment]
@alexkeizer Also, could you please update the README to describe the new `benchmark` target?

**Comment by bollu** [comment]
A separate build job makes sense to me, and PR LGTM!

**Comment by shigoel** [comment]
@alexkeizer FYI: incoming merge conflicts here.

**Comment by alexkeizer** [comment]
@shigoel I've added a line to the README, and made benchmarking part of the CI (on the ubuntu runner only, figured that'd be enough). This should now be ready to merge

**Review by shigoel — APPROVED** [review]
Thanks!

**Pull request merged**
